mountpoint-s3/src/inode.rs

```diff
@@ -354,6 +354,76 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Unlink the entry described by `parent_ino` and `name`.
+    ///
+    /// If the entry exists, delete it from S3 and the superblock.
+    /// 
+    /// We know that the Linux Kernel's VFS will lock both the parent and child,
+    /// so we can safely ignore concurrent operations within the same Mountpoint process to the file and its parent.
+    /// See: https://www.kernel.org/doc/html/next/filesystems/directory-locking.html
+    pub async fn unlink<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let parent = self.inner.get(parent_ino)?;
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::Directory {
+            return Err(InodeError::IsDirectory(inode.ino()));
+        }
+
+        // Take exclusive write locks early.
+        // We know that the VFS will already have locked its equivalent representations.
+        let mut parent_state = parent.inner.sync.write().unwrap();
+        // Take a write lock even if we don't mutate, since we would not want to permit any concurrent operations.
```

Comment:
Why are we taking write lock on child? Even if we take a read lock, it ensures there wont be any mutations. Others can only take read lock over it concurrently and not write lock.

Reply:
I agree - the main reason I took the write lock was just to be consistent with the VFS.

I've since removed this now.